### ISSUE TITLE
FN: `EqualsIncompatibleType` misses comparisons between unrelated record types

### ISSUE THREAD

**Issue by rincho0617** [google/error-prone#6107]
## What version of Error Prone are you using?

2.50.0

## Does this issue reproduce with the latest release?

Yes.

## What did you do?

```java
public class Test {
  record UserId(long id) {}

  record OrderId(long id) {}

  boolean f(UserId userId, OrderId orderId) {
    return userId.equals(orderId);
  }
}
```

## What did you expect to see?

`[EqualsIncompatibleType] Calling equals on incompatible types UserId and OrderId`

## What did you see instead?

No diagnostic. Comparisons where exactly one side is a record are flagged as expected,
so the gap is specific to record-to-record comparisons:

| comparison | flagged |
| --- | --- |
| `record` vs `java.util.UUID` | yes |
| `record` vs `enum` | yes |
| unrelated `final class` vs unrelated `final class` (both with a hand-written `equals`) | yes |
| `record` vs unrelated `record` | **no** |

`TypeCompatibility#isFeasiblyCompatible` treats two types as compatible when an
`equals(Object)` override on the left type is inherited by the right type.
`java.lang.Record` declares `public abstract boolean equals(Object)`, and
`customEqualsMethod` only excludes `Object` and `Enum`, so every pair of records
shares `Record.equals` and is considered compatible.

Like `Object.equals` and `Enum.equals`, `Record.equals` specifies no equality
semantics that two distinct record types could share, so it looks like it belongs
in the same exclusion. This is the same shape as #679 (enums) and #2384 (protos).

Records are increasingly used as wrapper types for IDs and other values, where
mixing two of them up is exactly the mistake this check is meant to catch.
